fix(r2): validate bucket names

vicb opened the pull request:
Fixes #9148

Before this PR, bucket names were only validated when creating a bucket (via `wrangler r2 create bucket_name`)

This PR adds validation to:
- wrangler config
- object commands - when trying to put, get, delete an object
- pages dev CLI command
- pipeline config

---

<!--
Please don't delete the checkboxes <3
The following selections do not need to be completed if this PR only contains changes to .md files
-->

- Tests
  - [ ] TODO (before merge)
  - [x] Tests included
  - [ ] Tests not necessary because:
- Wrangler / Vite E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [x] Required
  - [ ] Not required because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): <!--e.g. <https://github.com/cloudflare/cloudflare-docs/pull/>...-->
  - [x] Documentation not necessary because: names are documented
- Wrangler V3 Backport
  - [ ] TODO (before merge)
  - [x] Wrangler PR: https://github.com/cloudflare/workers-sdk/pull/9172
  - [ ] Not necessary because: patch

<!--
Have you read our [Contributing guide](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md)?
In particular, for non-trivial changes, please always engage on the issue or create a discussion or feature request issue first before writing your code.
-->

penalosa commented:
@vicb this needs a backport since it's a Wrangler patch

vicb commented:
> @vicb this needs a backport since it's a Wrangler patch

I don't think we want to backport that?

vicb commented:
> > @vicb this needs a backport since it's a Wrangler patch
> 
> I don't think we want to backport that?

Discussed offline, I'll create a v3 PR when this one is green and approved

petebacondarwin commented on fixtures/pages-functions-app/functions/r2/create.ts:
```diff
@@ -1,8 +1,8 @@
 type Env = {
-	BUCKET: R2Bucket;
+	bucket: R2Bucket;
```
Seems like an unnecessary change, or am I missing something?

vicb replied:
The name should be lowercase

petebacondarwin replied:
But this is not the bucket name, it is the binding name, which doesn't have the same validation requirements.

petebacondarwin commented on packages/wrangler/src/config/validation.ts:
```diff
@@ -2561,6 +2562,18 @@ const validateR2Binding: ValidatorFn = (diagnostics, field, value) => {
+		hasProperty(value, "bucket_name") &&
+		!isValidR2BucketName(value.bucket_name)
+	) {
+		diagnostics.errors.push(
+			`"${field}.bucket_name" must begin and end with an alphanumeric, only contain letters (a-z), numbers (0-9), and hyphens (-)," +
+			" and be between 3 and 63 characters long. but got ${JSON.stringify(value.bucket_name)}.`
+		);
```
This message is repeated six times in the code. Could it just be added to the `isValidR2BucketName()` function to avoid the duplication?

vicb replied:
We can do that, probably renaming the function to `validateR2BucketName` returning `{isValid, errorMsg?}`

petebacondarwin commented on packages/wrangler/src/pages/dev.ts:
```diff
@@ -1269,6 +1270,16 @@ function getBindingsFromArgs(args: typeof pagesDevCommand.args): Partial<
+				const bucketName = ref || binding.toString();
+
+				if (!isValidR2BucketName(bucketName)) {
+					logger.error(
+						`The bucket name "${bucketName}" is invalid. ` +
+							"Bucket names must begin and end with an alphanumeric, only contain letters (a-z), numbers (0-9), and hyphens (-), and be between 3 and 63 characters long."
+					);
+					return;
```
Why the quiet exit here?
If you throw a UserError then the process will exit with a non-zero exit code.

vicb replied:
Because this is similar to line 1270 and I don't any `UserError` here.

emily-shen commented on packages/wrangler/src/config/validation.ts:
```diff
@@ -2561,6 +2562,18 @@ const validateR2Binding: ValidatorFn = (diagnostics, field, value) => {
+		!isValidR2BucketName(value.bucket_name)
+	) {
+		diagnostics.errors.push(
+			`"${field}.bucket_name" must begin and end with an alphanumeric, only contain letters (a-z), numbers (0-9), and hyphens (-)," +
```
looks like uppercase letters are invalid too, even though the api error message gives zero indication about this :/

```suggestion
			`"${field}.bucket_name" must begin and end with an alphanumeric character, only contain lowercase letters (a-z), numbers (0-9), and hyphens (-)," +
```

emily-shen commented on packages/wrangler/src/pipelines/cli/create.ts:
```diff
@@ -157,6 +158,10 @@ export const pipelinesCreateCommand = createCommand({
 	async handler(args, { config }) {
 		const bucket = args.r2Bucket;
+		if (!isValidR2BucketName(bucket)) {
+			`The bucket name "${bucket}" is invalid. ` +
```
this doesn't seem to be throwing an error

vicb replied:
Good catch, it's missing a `throw UserError(...)` I guess

vicb commented:
@emily-shen @petebacondarwin thanks for the reviews, I have addressed the feedback, PTAL

- Factor and clarify the format message into a single const
- Add a missing `throw new UserError(...)`

emily-shen commented on packages/wrangler/src/config/validation.ts:
```diff
@@ -2569,6 +2581,17 @@ const validateR2Binding: ValidatorFn = (diagnostics, field, value) => {
 		isValid = false;
 	}
+	if (
+		isValid &&
```
with this `isValid` check, if you have invalid  bucket_name and invalid preview_bucket_name, only the error for bucket_name will show up I think?

vicb replied:
That's right.

I felt like this was good enough (i.e. you'll get the other error the next time if only one is fixed) vs making the code more complicated (to only validate when there is a name)

But if you feel strongly I can change that, LMK

emily-shen replied:
i think it we usually try our best to avoid that pattern - fixing something and exposing another error, although the error is the same here so its probably not a huge deal.

vicb replied:
I understand but at the same time I don't really understand why we have a `preview_bucket_name` when we have environments and I don't think other bindings have a similar thing.

arcticmatt commented:
would be nice if there was a flag/option to disable this behavior, since it's not backwards compatible with all `wrangler.jsonc` files that were working before this patch

edmundhung commented:
@arcticmatt We are relaxing R2 bucket validation for pages dev commands in #9257 based on the feedbacks and will release it very soon.

arcticmatt commented:
@edmundhung I'm not using pages dev

Here's an example of the issue I ran into:

```
pnpm wrangler d1 migrations apply DB_NAME --local

 ⛅️ wrangler 4.15.1
-------------------


✘ [ERROR] Processing wrangler.jsonc configuration:

    - r2_buckets[0].bucket_name="REPLACEME_R2_BUCKET_NAME" is invalid. Bucket names must begin and
  end with an alphanumeric character, only contain lowercase letters, numbers, and hyphens, and be
  between 3 and 63 characters long.
```

I have some existing code that uses all caps for bucket names, and then I replace the bucket name with a valid one once deploying. It's not too bad to start using lowercase for the placeholder name, but I didn't expect this to break when upgrading to a new minor version of Wrangler

Cherry commented:
Yeah, this being validated at wrangler config level, which happens for (almost?) every command in `wrangler` can be quite problematic when this wasn't previous enforced this way. Template or multi-stage-like projects where bindings are inserted at build time, etc. are going to suffer with errors now when upgrading.

If this is to be validated, I'd recommend it's only on operations where it fundamentally needs to be - like `r2` commands, deploying a worker, etc. Running local d1 migrations for example in a project with a `wrangler.json` that has a template like `<BINDING_NAME>` similar to the docs really shouldn't start causing commands to throw errors when upgrading to a new minor wrangler version.
